synthesis/synthesis.go

```diff
@@ -260,76 +178,60 @@ func FixCds(sqlitePath string, sequence string, codontable codon.Table, problema
-		tx.MustExec(`INSERT INTO history(pos, codon, step) VALUES (?, ?, 0)`, pos, codon)
-		pos++
+		aminoAcidWeightTable[aminoAcid.Letter] = aminoAcidTotal
 	}
 
+	// Build weight map
 	for _, aminoAcid := range codontable.AminoAcids {
 		for _, codon := range aminoAcid.Codons {
-			codonWeightRatio := float64(codon.Weight) / float64(weightTable[aminoAcid.Letter])
+			codonWeightRatio := float64(codon.Weight) / float64(aminoAcidWeightTable[aminoAcid.Letter])
 			normalizedCodonWeight := 100 * codonWeightRatio
-			tx.MustExec(`INSERT INTO weights(codon, weight) VALUES (?,?)`, codon.Triplet, normalizedCodonWeight)
+			weightMap[codon.Triplet] = normalizedCodonWeight
 		}
 	}
 
-	err := tx.Commit()
-	if err != nil {
-		return sequence, []Change{}, err
+	// Build historical map
+	pos := 0
+	for codonPosition := 0; codonPosition < len(sequence); codonPosition = codonPosition + codonLength {
+		codon := sequence[codonPosition : codonPosition+codonLength]
+		historicalMap[pos] = append(historicalMap[pos], codon)
+		pos++
 	}
 
 	// For a maximum of 100 iterations, see if we can do better. Usually sequences will be solved within 1-3 rounds,
 	// so 100 just effectively acts as the max cap for iterations. Once you get to 100, you pretty much know that
 	// we cannot fix the sequence.
-	for i := 1; i < fixIterations; i++ {
+	getSequence := func(history map[int][]string) string {
+		var sequence string
+		for codonPosition := 0; codonPosition < len(history); codonPosition++ {
+			sequence = sequence + history[codonPosition][len(history[codonPosition])-1]
+		}
+		return sequence
+	}
+	var changes []Change
+	for step := 1; step < fixIterations; step++ {
 		suggestions := findProblems(sequence, problematicSequenceFuncs)
 		// If there are no suggestions, break the iteration!
 		if len(suggestions) == 0 {
-			// Add a historical log of changes
-			var changes []Change
-			// This SQL will basically select positions + steps and organize
-			// the codons by firstly their positions, and then by the step. We
-			// want codons from the last fix iteration. It then selects the
-			// codons at each position that has been changed and returns what it
-			// has been changed to and what it has been changed from (this requires
-			// a subquery)
-			getChangesSQL := `SELECT h.pos             AS position,
-					       h.step            AS step,
-					       (SELECT codon
-					        FROM   history
-					        WHERE  pos = h.pos
-					               AND step = h.step - 1
-					        LIMIT  1)        AS codonfrom,
-					       h.codon           AS codonto,
-					       sf.suggestiontype AS reason
-					FROM   history AS h
-					       JOIN suggestedfix AS sf
-					         ON sf.id = h.suggestedfix
-					WHERE  h.suggestedfix IS NOT NULL
-					ORDER  BY h.step,
-					          h.pos 
-					`
-			_ = db.Select(&changes, getChangesSQL)
+			// Sort changes by step and position
+			sort.Slice(changes, func(i, j int) bool {
```

Comment:
Could you name these lamba variables some other than `i` and `j`?

Reply:
Idiomatic go https://pkg.go.dev/sort#Slice